programs/clearing_house/src/math/repeg.rs

```diff
@@ -428,6 +437,230 @@ pub fn adjust_peg_cost(
+) -> ClearingHouseResult<(Market, i128)> {
+    let curve_update_intensity = cast_to_i128(min(market.amm.curve_update_intensity, 100_u8))?;
+    // let current_peg = market.amm.peg_multiplier;
+
+    // return early
+    if optimal_peg == market.amm.peg_multiplier || curve_update_intensity == 0 {
+        // msg!("RETURN adjust_prepeg EARLY");
+        return Ok((*market, 0));
+    }
+
+    let delta_peg = cast_to_i128(optimal_peg)?
+        .checked_sub(cast_to_i128(market.amm.peg_multiplier)?)
+        .ok_or_else(math_error!())?;
+
+    let mut per_peg_cost = cast_to_i128(market.amm.quote_asset_reserve)?
+        .checked_sub(cast_to_i128(market.amm.terminal_quote_asset_reserve)?)
+        .ok_or_else(math_error!())?
+        .checked_div(AMM_TIMES_PEG_TO_QUOTE_PRECISION_RATIO_I128)
+        .ok_or_else(math_error!())?;
+
+    let budget_i128 = cast_to_i128(budget)?;
+
+    let mut market_clone = *market;
+    let mut budget_delta_peg: i128 = 0;
+    let mut budget_delta_peg_magnitude: u128 = 0;
+    let cost: i128;
+    let new_peg: u128;
+
+    if per_peg_cost != 0 {
+        budget_delta_peg = budget_i128
+            // .checked_mul(AMM_TIMES_PEG_TO_QUOTE_PRECISION_RATIO_I128)
+            // .ok_or_else(math_error!())?
+            .checked_div(per_peg_cost)
+            .ok_or_else(math_error!())?;
+        budget_delta_peg_magnitude = budget_delta_peg.unsigned_abs();
+    }
+
+    if (per_peg_cost == 0
+        || budget_delta_peg > 0 && delta_peg < 0
+        || budget_delta_peg < 0 && delta_peg > 0)
+        || (budget_delta_peg_magnitude > delta_peg.unsigned_abs())
+    {
+        // use optimal peg
+        cost = per_peg_cost
+            .checked_mul(delta_peg)
+            .ok_or_else(math_error!())?;
+        // .checked_div(AMM_TIMES_PEG_TO_QUOTE_PRECISION_RATIO_I128)
+        // .ok_or_else(math_error!())?;
+        new_peg = optimal_peg;
+    } else {
+        // use full budget peg
+
+        // equivalent to (but cheaper than) scaling down by .1%
+        // let (k_scale_numerator, k_scale_denominator) = (999, 1000);
+        let adjustment_cost: i128 = if adjust_k {
+            // TODO can be off by 1?
+
+            let new_sqrt_k = market
+                .amm
+                .sqrt_k
+                .checked_sub(
+                    market
+                        .amm
+                        .sqrt_k
+                        .checked_div(1000)
+                        .ok_or_else(math_error!())?,
+                )
+                .ok_or_else(math_error!())?;
+
+            let new_base_asset_reserve = market
+                .amm
+                .base_asset_reserve
+                .checked_sub(
+                    market
+                        .amm
+                        .base_asset_reserve
+                        .checked_div(1000)
+                        .ok_or_else(math_error!())?,
+                )
+                .ok_or_else(math_error!())?;
+            let new_quote_asset_reserve = market
```

Comment:
why dont you divide invariant by new base asset reserve to get this? this leads to the rounding errors we've seen before?